### ISSUE TITLE
Labels cache file takes name of random folder in dataset

### ISSUE THREAD

**Issue by RyanHir** [ultralytics/yolov5#8384]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

Using the instructions from https://community.ultralytics.com/t/how-to-combine-weights-to-detect-from-multiple-datasets/38/5 to create a `yaml` with the syntax to allow for multipath reading, I was expecting to have `detect.py` read the contents of each directory to generate the label cache file (and upload the dataset to memory when `--cache` is used).

However, it appears that only one folder is used in both the train and validate datasets. In the screenshot below, you can see that only `arvalis_1` and `ethz_1` is loaded. This config is the GlobalWheat2020 project located at https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml.

![image](https://user-images.githubusercontent.com/4690732/176300006-92854da4-3704-41b2-9344-f00cacf8b121.png)

### Environment

- YOLOv5 🚀 v6.1-269-gf76a78e torch-1.11.0+cu102
  - CUDA:0 (NVIDIA GeForce GTX 1080 Ti, 11178MiB)
  - CUDA:1 (NVIDIA GeForce GTX 1080 Ti, 11178MiB)
- OS: Ubuntu 20.04
- Python: 3.8.10

### Minimal Reproducible Example

`python -m torch.distributed.run --nproc_per_node 2 train.py --img 640 --batch 128 --data data/GlobalWheat2020.yaml --weights yolov5n.pt --device 0,1 --epochs 30 --name global_wheat`

but is also reproducible with one GPU

`python train.py --img 640 --batch 64 --data data/GlobalWheat2020.yaml --weights yolov5n.pt --device 0 --epochs 30 --name global_wheat`

### Additional

This issue was initially found with a config similar to the following
```yaml
path: /mnt/disk1/data/Datasets/training
train:
 - images/orig/train
 - images/augment/train
val:
 - images/orig/valid
 - images/augment/valid
nc: 3
names: ['a' 'b', 'c']
```

Folder structure looks like...
```
/mnt/disk1/data/Datasets/training
├── dataset.yaml
├── images
│   ├── augment
│   │   ├── train
│   │   └── valid
│   └── orig
│       ├── train
│       └── valid
├── labels
    ├── augment
    │   ├── train
    │   ├── train.cache
    │   ├── valid
    │   └── valid.cache
    └── orig
        ├── train
        └── valid
```

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@RyanHir thanks for the bug report! Can you confirm that GlobalWheat uses all images without --cache but only uses one directory when --cache is applied?

**Comment by RyanHir** [comment]
@glenn-jocher This is occurring with and without the --cache flag.

**Comment by glenn-jocher** [comment]
@RyanHir cache and console readouts are probably indicating last or first directory only when printing to console. How many images are in GlobalWheat in total compared to tqdm progress bar?

**Comment by glenn-jocher** [comment]
@RyanHir looks like dataset contains 3422 images in total, same as tqdm from your screenshot. Is that correct?

<img width="1419" alt="Screen Shot 2022-06-29 at 5 08 32 PM" src="https://user-images.githubusercontent.com/26833433/176471348-cfe26c08-9167-452c-bd22-63596c48df91.png">

**Comment by RyanHir** [comment]
@glenn-jocher That looks correct on my end. It appears that even my own dataset is caching all the labels in after comparing the number of files I have versus what tqdm is showing.

So I now believe this is only a visual bug. Perhaps the label cache file that is saved on disk can be given a more proper name?

**Comment by RyanHir** [comment]
Ideally this behavior is changed, as I have a use case for having different dataset yaml files in the same directory, referencing the same data. So having the collisions can cause a headache as the system rebuilds the caches.

**Comment by RyanHir** [comment]
> @RyanHir cache and console readouts are probably indicating last or first directory only when printing to console. How many images are in GlobalWheat in total compared to tqdm progress bar?

@glenn-jocher I did notice today that the choice of naming did change after doing a rerun. However, I am not sure how to reproduce this.

**Comment by glenn-jocher** [comment]
@RyanHir got it. Right now the logic for naming `cache_path` is here on L449. If there's a more robust method for your use case then please submit a PR to help integrate these improvements most efficiently. Thank you!

https://github.com/ultralytics/yolov5/blob/e50dc38d3687d18cd932aa342bca03ca7125bbe0/utils/dataloaders.py#L447-L456

**Comment by RyanHir** [comment]
Thank you! I will take a look

**Comment by glenn-jocher** [comment]
@RyanHir you're welcome! Feel free to reach out if you have any questions or need further assistance. Thank you for your contribution to the YOLO community!
